pageserver/src/layered_repository.rs

```diff
@@ -0,0 +1,1056 @@
+    fn advance_last_valid_lsn(&self, lsn: Lsn) {
+        let old = self.last_valid_lsn.advance(lsn);
+
+        // Can't move backwards.
+        if lsn < old {
+            warn!(
+                "attempted to move last valid LSN backwards (was {}, new {})",
+                old, lsn
+            );
+        }
+    }
+
+    fn init_valid_lsn(&self, lsn: Lsn) {
+        let old = self.last_valid_lsn.advance(lsn);
+        assert!(old == Lsn(0));
+        let old = self.last_record_lsn.fetch_max(lsn);
+        assert!(old == Lsn(0));
+        self.prev_record_lsn.store(Lsn(0));
+    }
+
+    fn get_last_valid_lsn(&self) -> Lsn {
+        self.last_valid_lsn.load()
+    }
+
+    ///
+    /// Remember the (end of) last valid WAL record remembered in the page cache.
+    ///
+    /// NOTE: this updates last_valid_lsn as well.
+    ///
+    fn advance_last_record_lsn(&self, lsn: Lsn) {
+        // Can't move backwards.
+        let old = self.last_record_lsn.fetch_max(lsn);
+        assert!(old <= lsn);
+
+        // Use old value of last_record_lsn as prev_record_lsn
+        self.prev_record_lsn.fetch_max(old);
+
+        // Also advance last_valid_lsn
+        let old = self.last_valid_lsn.advance(lsn);
+        // Can't move backwards.
+        if lsn < old {
+            warn!(
+                "attempted to move last record LSN backwards (was {}, new {})",
+                old, lsn
+            );
+        }
+    }
+
+    fn get_last_record_lsn(&self) -> Lsn {
+        self.last_record_lsn.load()
+    }
+
+    fn get_prev_record_lsn(&self) -> Lsn {
+        self.prev_record_lsn.load()
+    }
+}
+
+impl LayeredTimeline {
+    /// Open a Timeline handle.
+    ///
+    /// Loads the metadata for the timeline into memory.
+    fn new(
+        conf: &'static PageServerConf,
+        metadata: TimelineMetadata,
+        ancestor: Option<Arc<LayeredTimeline>>,
+        timelineid: ZTimelineId,
+        tenantid: ZTenantId,
+        walredo_mgr: Arc<dyn WalRedoManager + Send + Sync>,
+    ) -> Result<LayeredTimeline> {
+        let timeline = LayeredTimeline {
+            conf,
+            timelineid,
+            tenantid,
+            layers: Mutex::new(LayerMap::default()),
+
+            walredo_mgr,
+
+            last_valid_lsn: SeqWait::new(metadata.last_valid_lsn),
+            last_record_lsn: AtomicLsn::new(metadata.last_record_lsn.0),
+            prev_record_lsn: AtomicLsn::new(metadata.prev_record_lsn.0),
+
+            ancestor_timeline: ancestor,
+            ancestor_lsn: metadata.ancestor_lsn,
+        };
+        Ok(timeline)
+    }
+
+    ///
+    /// Get a handle to a Layer for reading.
+    ///
+    /// The returned SnapshotFile might be from an ancestor timeline, if the
+    /// relation hasn't been updated on this timeline yet.
+    ///
+    fn get_layer_for_read(
+        &self,
+        rel: RelishTag,
+        lsn: Lsn,
+    ) -> Result<Option<(Arc<dyn Layer>, Lsn)>> {
+        // First dig the right ancestor timeline
```

Comment:
What does this comment refer to?

Reply:
It actually refers to the code a few lines below. Moved things around to make it more clear.